CMakeLists.txt

```diff
@@ -377,8 +382,13 @@ if (BUILD_RDIFF)
   else (POPT_FOUND)
     message (WARNING "Popt library is required for rdiff target")
   endif (POPT_FOUND)
+  target_compile_options(rdiff PRIVATE -DLIBRSYNC_STATIC_DEFINE)
```

Comment:
We don't want rdiff statically linked, we want it to dynamically link against the rsync.dll or librsync.so library.

So we don't want this line. However, I'm curious to know if you found you needed this for some reason under MinGW and/or MSVC. If you do need it for some reason, I suspect there is another problem we need to fix for MinGW and/or MSVC on win32 platforms.

Reply:
Fixed. Without the static define, rdiff failed to compile first due to "initializer element is not constant" for `rs_inbuflen` and `rs_outbuflen` (which are defined in `whole.c`), and then when I added second definitions for them in `rdiff.c`, a `function is marked dllimport` error resulted due to the LIBRSYNC_EXPORT qualifier. Ultimately, I fixed this by removing the `rs_*` prefixes and the qualifier from both variables in rdiff.c, making them distinct from the variables declared in `librsync.h` and defined in `whole.c`.

Reply:
Sorry, I didn't see this comment until just now... it was hidden as "resolved". Defining new variables in rdiff.c is definitely wrong, but I'm not sure why you hit the problems you did.

That error message is strange, and doesn't seem to correspond to what the online searches for that error say it's about. I have a sneaking suspicion that the compiler is getting confused because we have 2 variables exported/defined on the same line with a single LIBRSYNC_EXPORT for them both.

Perhaps edit whole.c AND librsync.h and put them each on their own line with their own LIBRSYNC_EXPORT, and see if that works.

Another thing that might be worth trying is adding "static" to their definition, though that shouldn't be needed.

Reply:
Doesn't work. The `dllimport` attribute may actually [be the reason](http://mingw.5.n7.nabble.com/error-with-a-function-declared-with-dllimport-and-used-in-a-struct-td13576.html) for the `initializer element is not constant` error.

Reply:
I found similar docs about "initializer element is not constant error.", but the explanations didn't seem to match the situation we have here; rs_inbuflen and rs_outbuflen are not constants, and they are being initialized to constant '0' values in whole.c.

However, I think I now understand why this is happening. The problem is not strictly rs_inbuflen or rs_outbuflen, it's "const struct poptOption opts[]" in rdiff.c that we are initializing with a value that includes '&rs_inbuflen" and "&rs_outbuflen" which are unknown at compile time, since it's only after linking that their addresses are known. This appears to be a limitation of dll's, and linux's dynamic linking doesn't seem to have this limitation.

To fix this is a bit tricky. I think probably the neatest way is to move the whole opts definition inside main() in rdiff.c, so that it is initialized at runtime. Other options include declaring local int's for inbuflen and outbuflen to put into opts (like you did before) and make rdiff_options() set rs_inbufflen = inbuflen and rs_outbuflen=outbuflen after all the arguments are parsed to make sure the cmdline arguments are injected into whole.c where they will be used (which you didn't have before). I think making opts be initialized at runtime is neater though, and we should perhaps add a comment explaining why.

I've also figured out why rs_librsync_version was not causing problems without LIBRSYNC_EXPORT defined on it in librsync.h; it's because version.c where it is defined doesn't import librsync.h, and thus the compiler doesn't see the miss-matched declaration/definition when it's compiling version.c. However, for the others whole.c and sumset.c DO import librsync.h, so it's important that the declarations in librsync.h match the definitions in the *,c files.

The main thing that doesn't make sense to me so far is that it appeared to work with LIBRSYNC_STATIC_DEFINE set for compiling rdiff. Everything I've read suggests that symbols imported from DLLs need to have __declspec(dllimport) defined on them to work. For it to appear to work without it suggests that it was somehow statically linking rdiff, but the way the cmake stuff is setup I can't see how it would have been doing that.

To be strictly correct, rs_librsync_version should have LIBRSYNC_EXPORT defined on it in librsync.h too. I see you've done this... thanks.

I think we've finally figured out how to make this work.

Reply:
Fixed. Thanks for the help!